### PR TITLE
Add support for T320

### PR DESCRIPTION
**Pull request by StarF666**
Had to rework FW file parsing to detect firmware for the T320. Added an additional integrity check (FW seems to have a 256 meta header). My T320 immediatley resets connection after entering bootloader or finishing update. Therefore exception handling was added and console/debug messages have been adapted. Update to 1.12 works on my T320. Hope this helps anybody. Thanks a lot to c0d3z3r0 for doing all the inital reverse engineering.

### CODE REVIEW

File: t3xupdate.py [hunk 1]
```diff
@@ -71,9 +71,17 @@ def parse_update(self, file):
         if fw_size < 0x100:
             error(f"Bad update image file size: {fw_size} bytes.")
             return (None, None, 0)
+        
+        self.file.seek(0x60)
+        fw_payload_size = int.from_bytes(self.file.read(4), "big")
+        fw_payload_diff = fw_size - fw_payload_size
+        if fw_payload_diff != 256:
```
Comment (c0d3z3r0) [inline t3xupdate.py]: Yep, good idea to check that! let's do it the other way round, though, for clarity:
1. rename `fw_size` to `file_size`
2. drop lines 71-73
3.
```python
self.file.seek(0x60)
fw_size = int.from_bytes(self.file.read(4), "big") + 0x100
if file_size != fw_size:
  error("Bad firmware file size: Expected {fw_size} bytes but file is {file_size} bytes.")
  return (None, None, 0)
```
Also, this could be a separate commit

File: t3xupdate.py [hunk 2]
```diff
@@ -71,9 +71,17 @@ def parse_update(self, file):
...
+            error(f"Bad image file: Payload difference is {fw_payload_diff} bytes. Should be 256 bytes!")
+            return (None, None, 0)
+
 
         self.file.seek(0x20)
-        fw_product = self.file.read(8).split(b'_')[-1]
+        fw_product = self.file.read(16).split(b'\xff')[0].split(b'_')[-1]
```
Comment (c0d3z3r0) [inline t3xupdate.py]: use `rstrip(b'\xff')` instead, please

File: t3xupdate.py [hunk 3]
```diff
@@ -158,7 +180,7 @@ def do_update(self, file):
             error("Update failed!")
             return False
 
-        info("Update successful!")
+        info("Update successful! Installed FW version {fw_version}")
```
Comment (c0d3z3r0) [inline t3xupdate.py]: Could be a separate commit

**Comment by c0d3z3r0** [comment]
> Had to rework FW file parsing to detect firmware for the T320. Added an additional integrity check (FW seems to have a 256 meta header). My T320 immediatley resets connection after entering bootloader or finishing update. Therefore exception handling was added and console/debug messages have been adapted. Update to 1.12 works on my T320. Hope this helps anybody. Thanks a lot to c0d3z3r0 for doing all the inital reverse engineering.

Awesome! Thank you very much for your contribution! :-)

You are right about the fw header. It's the same for T3A. I just realized there is a CRC16 in there, too, which can be checked. Will push it later.

I wonder, if we can simply the update portion of your change. Could you try the version from https://github.com/c0d3z3r0/aixun_t3x_updater/tree/test on your T320, please? I would expect that stopping sending after the max. payload size is enough to prevent that failure.

The next step would be to check the firmware version after update to verify everything went well. (Oops, I already did that haha)

**EDIT:** Oh, no... that's not the point. The problem seems to be that the T320 isn't expecting the last packet of 2048 bytes. Could you flash a older firmware and then do an update with the original Aixun tool, while running Wireshark+usbmon? Maybe that helps finding out what is wrong.

**EDIT 2**: Uhm, well, I had another look at the T3A bootloader code. Maybe... just maybe... the firmware size in the header includes the header size, and we just need to skip the last 256 bytes... I've pushed another version to https://github.com/c0d3z3r0/aixun_t3x_updater/tree/test for you to test, if you like

**Review by c0d3z3r0 — COMMENTED** [review]
I would split this into four commits:
1. Verify payload size
2. Support longer product name string
3. Improve bootloader error handling to support e.g. T320
4. Print firmware version after update

File: t3xupdate.py (hunk 2 above)
Comment (c0d3z3r0) [reply t3xupdate.py]: Could be a separate commit as well. "Support longer product name string"

**Comment by StarF666** [comment]
Thank you for your work. I tried the test branch but it only works with a chunk size of 2048, anything else fails. Smaller values don't respond with 'ack_jcxx', 4096 responded with 'ack_jcxxack_jcxx' and broke the code (if condition fired). 

The display of the station also shows percentage of update so I wonder if the bootloader expects a specific firmware size (how else could it calculate percentage?). 

I'll split up my commits like you proposed. Sorry for my sloppiness, I'm not working as a developer for quite some time now and all this reverse enginieering is new to me. Just wanted to fix my tube amp with my new soldering station an now I'm recompiling MCU firmware. Thanks for that. :) 

Gonna setup wireshark and usbmon. The T320 has a esp32 network chip and the firmware has two commands "JC_to_esp32:" and "JC_from_esp32:" probably for UART communication to the esp32. I'd like to see if it is possible to set wifi without the Aixun software but that's probably easier through monitoring communication with the station.

**Comment by c0d3z3r0** [comment]
> Thank you for your work. I tried the test branch but it only works with a chunk size of 2048, anything else fails. Smaller values don't respond with 'ack_jcxx', 4096 responded with 'ack_jcxxack_jcxx' and broke the code (if condition fired).

Ah, yes, ofc I missed that `ack_jcxx` is only sent after each 2048 bytes. I'll wait for the usbmon log, to see what the vendor tool does.

This is how T3A bootloader looks like:

```c
int __fastcall usb_process_data(unsigned int *data, unsigned int len)
{
  unsigned __int8 *out_len; // r0 MAPDST
  char *tmp_buf; // [sp+0h] [bp-18h] BYREF

  if ( do_update == 1 )
  {
    if ( do_usb_firmware_update(data, len) == 1 )// ack each 2048 bytes
    {
      strcpy(out_data, "ack_jcxx");
      out_len = strlen(out_data);
      copy_usb_out_data(out_data, out_len);
    }

    return 1;
  }

  else if ( !strncmp(data, "jcxxstid", 8u) )
  {
    *out_data = CHIPID[0];
    *&out_data[4] = CHIPID[1];
    *&out_data[8] = CHIPID[2];
    copy_usb_out_data(out_data, 0xCu);
    return 1;
  }

  else if ( !strncmp(data, "JC_identity", 0xBu) )
  {
    strcpy(out_data, "JC_boot");
    out_len = strlen(out_data);
    copy_usb_out_data(out_data, out_len);
    return 1;
  }

  else if ( !strncmp(data, "JC_version", 0xAu) )
  {
    sprintf(out_data, "%s%s", "00160000", "JC_M_T3A_version1.01");
    copy_usb_out_data(out_data, 40u);
    return 1;
  }

  else if ( strncmp(data, "0x", 2u) || !strcmp(data, "JC_M_T3A_version1.01") )
  {
    if ( !strncmp(data, "JC_end", 6u) )
    {
      if ( fw_info.image_status.program_image_signature_success == 1 )
        strcpy(out_data, "end_jcxx");
      else
        strcpy(out_data, "fail_jcxx");

      out_len = strlen(out_data);
      copy_usb_out_data(out_data, out_len);
      return 1;
    }
    else
    {
      strncmp(data, "JC_end", 6u);
      return 1;
    }
  }
  else
  {
    fwupdate_buf_len = 0;
    mainfw_offset = 0;
    data_len_for_ack = 0;
    fwupdate_size = strtoull(data + 2, &tmp_buf, 16);
    do_update = 1;
    strcpy(out_data, "update_jcxx");
    out_len = strlen(out_data);
    copy_usb_out_data(out_data, out_len);
    return 1;
  }
}

int __fastcall do_usb_firmware_update(unsigned int *data, unsigned int len)
{
  memcpy(&fwupdate_buf + fwupdate_buf_len, data, len);
  fwupdate_buf_len += len;
  data_len_for_ack += len;
  if ( mainfw_offset || fwupdate_buf_len <= 255 )
  {
    if ( fwupdate_buf_len < 2048 )
    {
      if ( mainfw_offset + fwupdate_buf_len >= fwupdate_size )
      {
        flash_program_fw(&fwupdate_buf, &fw_info, mainfw_offset - 256, fwupdate_buf_len);
        mainfw_offset += fwupdate_buf_len;
        fwupdate_buf_len = 0;
        flash_image_signature();
        data_len_for_ack = 0;
        do_update = 0;
        return 1;
      }
    }
    else
    {
      flash_program_fw(&fwupdate_buf, &fw_info, mainfw_offset - 256, 2048u);
      fwupdate_buf_len -= 2048;
      memcpy(&fwupdate_buf, &unk_20000E78, fwupdate_buf_len);
      mainfw_offset += 2048;
      if ( mainfw_offset >= fwupdate_size )
      {
        flash_image_signature();
        data_len_for_ack = 0;
        do_update = 0;
        return 1;
      }
    }
  }
  else
  {
    validate_sent_fw_info(&fwupdate_buf, &fw_info);
    fwupdate_buf_len -= 256;
    memcpy(&fwupdate_buf, &unk_20000778, fwupdate_buf_len);
    mainfw_offset += 256;
  }

  if ( data_len_for_ack < 2048 )
    return 0;

  data_len_for_ack -= 2048;
  return 1;
}
```

> 
> The display of the station also shows percentage of update so I wonder if the bootloader expects a specific firmware size (how else could it calculate percentage?).

The firmware size is sent to the station. https://github.com/c0d3z3r0/aixun_t3x_updater/blob/master/t3xupdate.py#L138C1-L139C1

> 
> I'll split up my commits like you proposed. Sorry for my sloppiness, I'm not working as a developer for quite some time now and all this reverse enginieering is new to me.

No worries, I'm happy to help :-) Looks like something went wrong when pushing? There's still just one commit (and a merge commit).

> Gonna setup wireshark and usbmon. The T320 has a esp32 network chip and the firmware has two commands "JC_to_esp32:" and "JC_from_esp32:" probably for UART communication to the esp32. I'd like to see if it is possible to set wifi without the Aixun software but that's probably easier through monitoring communication with the station.

\o/ Maybe you can try to dump the firmware from the esp32?

**Comment by StarF666** [comment]
Yeah, I realized, firmware size is communicated to the bootloader. Next up is a new pull request. Shall we continue discussion there? Or open up an issue? I might need some help with wireshark/usbmon. And thanks for the code for the bootloader. I played around with Ghidra but this reverse engineering stuff is alien tech to me. Firmware dump for the esp32 would be nice. There is even a release for esp32 firmware on the Aixun repository.

**Comment by c0d3z3r0** [comment]
Let's continue here to keep information in one place.

Regarding Wireshark, I never used it on Windows directly to capture USB traffic, but I always capture on Linux while running a Windows VM (KVM/qemu). But IIRC usbpcap is included in the Wireshark installer. You will have to find out on which controller your station is connected and select the right one in Wireshark. (probably trial&error while reconnecting will work for you)

That should help: https://www.youtube.com/watch?v=3jpoUFUsui8
Note that it won't be possible to just select the device because the station will disconnected twice (reboot to bootloader and after flashing) and be detected as new device each time.

**Comment by StarF666** [comment]
Got wireshark running and tested a bit more. First, a youtube video about the T320 showing of the guts:

https://youtu.be/-ug7rLdOWto?t=752

There is a programming header for the esp32 but looking at the video it might just be parallel to the connection from the MCU. As the command for communication to the esp32 is "JC_to_esp32:" and by the looks of the traffic, it might be the same interface so maybe we can get to the firmware.

EDIT: Not so sure about the programming header, the esp32 has all sorts of connections through GPIOs. It supports UART, JTAG, USB, USB/Serial and SPI interfaces, all very flexible routable to the GPIOs. 

I captured the communication between the Aixun software and the station. When loading up the software, the following calls are being made to the station, with <- return Value:

JC_identity <- JC_User
JC_hardware <- JCUser
JC_display_name <-
date <-
JC_version <- JC_M_T320_version1.12
JCXX_open <- !!12666\x05L496keji
JC_to_esp32:\x01\x00\t\x00\x00\x00\x01P\xd7\r\n <- JC_from_esp32:\x01\xaa\x0e\x00\x00\x00\x01P\x041.10\xbd\r\n

I made a small debug tool from your updater and sent these commands manually and they return the same values.  "JC_display_name" and "date" don't return anything and I don't find those strings in the firmware. To get anything in return from the "JC_to_esp32:..." call, I have to call JC_identity and JCXX_open first. The return value of the esp32 seems to contain a version number. There are some subsequent calls returning the available network ids but alway encapsulated in some (to me unreadable) hex code. I tried to match this with some of the esp32 documentation. Could this be some kind of UART code? Or is the JCXX_open return value some kind of key/salt? But then I could'n read the SSIDs in the subsequent communication.

EDIT: See above, UART is just one of the possible connections. MCU is GD32F305RCT6. I tried to mix and match possible connections from both datasheets. But I don't realy know what I'm doing here.

These are my findings so far. Any idea, what JCXX_open is for? And how to decode the esp32 communication?

**Comment by c0d3z3r0** [comment]
Nice! Can you try to capture one update flow and upload the capture file? Then I can see, if the T320 update flow can be simplified a bit in #3.

`JCXX_open` returns `<CHIPID>keji`, see code below.

Regarding esp32 communication: I expect that the GD32 firmware decodes the commands from the Aixun software and forwards them to the esp32 and the other way round. You can either sniff the UART connecting using a logic analyzer (some FTDI-based one should work) or reverse engineer the esp32 firmware.

From T3A main firmware v1.26

```c
int __fastcall usb_process_data(unsigned int *data, unsigned int len)
{
  unsigned __int8 *v7; // r3
  unsigned __int8 *out_len; // r0 MAPDST

  if ( !strncmp((int)data, (int)"JC_image_end", 0xCu) )
  {
    data_len_for_ack = 0;
    do_image_flash = 0;
    usb_disable();
    delay_1ms(100);
    NVIC_SystemReset();
  }

  if ( do_image_flash )
  {
    if ( do_usb_bootimage_flash(data, len) == 1 )   // ack each 2048 bytes
    {
      strcpy(out_data, "image_ack");
      out_len = strlen(out_data);
      copy_usb_out_data((unsigned int *)out_data, (unsigned int)out_len);
    }

    return 1;
  }

  else if ( !strncmp((int)data, (int)"JCXX_open", 9u) )
  {
    *(_DWORD *)out_data = CHIPID[0];
    *(_DWORD *)&out_data[4] = CHIPID[1];
    *(_DWORD *)&out_data[8] = CHIPID[2];
    strcpy(&out_data[12], "keji");
    copy_usb_out_data((unsigned int *)out_data, 16u);
    return 1;
  }

  else if ( !strncmp((int)data, (int)"JC_identity", 11u) )
  {
    strcpy(out_data, "JC_User");
    out_len = strlen(out_data);
    copy_usb_out_data((unsigned int *)out_data, (unsigned int)out_len);
    set_jc_identity(1);
    return 1;
  }

  else if ( !strncmp((int)data, (int)"JC_version", 10u) )
  {
    sprintf((int)out_data, "%s", "JC_M_T3A_version1.26", v7);
    out_len = strlen(out_data);
    copy_usb_out_data((unsigned int *)out_data, (unsigned int)out_len);
    return 1;
  }

  else if ( !strncmp((int)data, (int)"JC_M_T3A_version1.26", 0x14u) )
  {
    strcpy(out_data, "JC_reset");
    out_len = strlen(out_data);
    copy_usb_out_data((unsigned int *)out_data, (unsigned int)out_len);
    jc_reset();
    return 1;
  }

  else if ( !strncmp((int)data, (int)"syscfg -v\r\nJC RW\r\n", 0x12u) )
  {
    strcpy(out_data, "JC_M_T3A");
    out_len = strlen(out_data);
    copy_usb_out_data((unsigned int *)out_data, (unsigned int)out_len);
    return 1;
  }

  else if ( !strncmp((int)data, (int)"JC DFU\r\n", 8u) )
  {
    return 1;
  }

  else if ( !strncmp((int)data, (int)"JC_image", 8u) && !strncmp((int)(data + 6), (int)"end", 3u) )
  {
    if ( check_boot_image_len((uint8_t *)data, len) )
    {
      image_buf_len = 0;
      image_offset = 0;
      data_len_for_ack = 0;
      do_image_flash = 1;
      strcpy(out_data, "image_start");
    }
    else
    {
      strcpy(out_data, "image_restart");
    }

    out_len = strlen(out_data);
    copy_usb_out_data((unsigned int *)out_data, (unsigned int)out_len);
    return 1;
  }
  else
  {
    return 1;
  }
}
```

**Comment by StarF666** [comment]
Nice code. Did you do that with IDA pro? Could you have a look at the T320 firmware what JC_to_eps32: actually does?

Here is the captured communication of the Aixun tool (v19.15.0) update procedure. Let me know if you need anything else.

[T320 update.pcapng.gz](https://github.com/user-attachments/files/15732592/T320.update.pcapng.gz)

Honestly, I flashed the station more than I used it for soldering. ;) But this is fun.

**Comment by c0d3z3r0** [comment]
> Nice code. Did you do that with IDA pro? Could you have a look at the T320 firmware what JC_to_eps32: actually does?

Yep, IDA Pro. Already on it :-)

> Here is the captured communication of the Aixun tool (v19.15.0) update procedure. Let me know if you need anything else.
> 
> [T320 update.pcapng.gz](https://github.com/user-attachments/files/15732592/T320.update.pcapng.gz)

Thanks!

> 
> Honestly, I flashed the station more than I used it for soldering. ;) But this is fun.

:laughing:

**Comment by c0d3z3r0** [comment]
Aha! They use Segger emWin für the GUI... I bet they used it for the T3A as well

image

**Comment by c0d3z3r0** [comment]
Ok, the T320 indeed does not send an ACK after the last 2048 byte package......

image

**Comment by c0d3z3r0** [comment]
Looks like the STM32 simply forwards anything after `JC_to_esp32:` to the esp32, reads the response and returns it as `JC_from_esp32:<response>`

```c
  else if ( !strncmp(data, "JC_to_esp32:", 0xCu) )
  {
    if ( len > 10 )
    {
      v13 = sub_8027A74(len);
      memcpy(v13, (data + 3), len - 14);
      write_esp32_uart(v13, len - 14);
      MEMORY[0x20000174] = 1;
      sub_802630C(v13);
    }

    return 1;
  }
  else
  {
```

Are you sure you bought a solder station, not a weather station? \o/

image

**Comment by StarF666** [comment]
Yeah, I guess they thought, let's do something more while we're at it (having network). 

Any idea, what protocol is used for communication with the esp32? Doesn't look like you could simply pass commands/arguments through the JC_to_esp32: function. At least it needs some encoding.

**Comment by c0d3z3r0** [comment]
No idea tbh, might be some cmd id + data enconding. I guess playing around with Wireshark and comparing packets should help
